#6 created a Client-side functionality for the Helpful and Not Helpfu…

Yjohn opened the pull request:
#6 it needs more work to make the buttons functional

Yjohn committed:
#6 created a Client-side functionality for the Helpful and Not Helpful buttons

Yjohn committed:
#6 created a Client-side functionality for the Helpful and Not Helpful buttons

NateWr commented on views/index.hbs:
```diff
@@ -8,23 +8,23 @@
+		<div class="entry__help">
+			<span id="displayCount">{{this.helpful}}</span>/<span id="totalNumber"></span> people found this helpful. Was this entry helpful?
+			<button id="Button1" class="button button--small">Yes</button>
+			<button id="Button2" class="button button--small">No</button>
```
`Button1` and `Button2` are bad names for the buttons. Consider using a more descriptive name like `button-helpful` and `button-not-helpful`.

NateWr commented on public/main.js:
```diff
@@ -0,0 +1,17 @@
+    unhelpful++;
+    total++;
+    totalNumber.innerHTML = total;
+});
```
You're going to run into problems using `getElementById`, because every `id` on an HTML page must be completely unique. You can't have two elements with the same `id`.

Learn about [querySelectorAll](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelectorAll) which allows you to select by class name and other attributes.

You'll face the same problem with your global `helpful` and `total` vars. If each question has it's own values, you'll need a way to store those values separately.

NateWr commented on views/layout.hbs:
```diff
@@ -14,5 +14,6 @@
 			</header>
 			{{{body}}}
 		</div>
+    <script src="/main.js"></script>
```
The indentation doesn't look right here. Try formatting the document.

NateWr reviewed (changes requested):
Good start! Just a few things to touch up and then start working on the problem of how you'll get each question's button to work separately.

Commit:
Add the rating route

Commit:
Take the FS from the code

Commit:
added the issues number #7

Yjohn committed:
#6 creating client side functionality, listen all the buttons ...

sending AJAX request to the route called 'rating' then resiving the change and display all the changes

Yjohn committed:
#6 creating client side functionality, listen all the buttons ...

sending AJAX request to the route called 'rating' then resiving the change and display all the changes

Yjohn committed:
fixing the function of receiveResponse

Yjohn committed:
change the name of not helpful to unhelpful

Yjohn committed:
change from string to validate JSON object by using JSON.parse

Yjohn committed:
fixing the merge conflict

Yjohn commented:
I saw the review and changed a lot of staff

NateWr commented on public/main.js:
```diff
@@ -0,0 +1,32 @@
+            helpful: helpful
+        })
+    };
+    fetch("/rating", options)
```
Where does the `fetch` function come from?

Yjohn replied:
fetch is built-in in modern browsers (i can use fetch without using Jquery or any other library)

Yjohn replied:
https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch

NateWr replied:
Cool! I didn't even know about that!

NateWr commented on public/main.js:
```diff
@@ -0,0 +1,32 @@
+        })
+    };
+    fetch("/rating", options)
+        .then(response => response.json())
```
What's this line doing?

Yjohn replied:
the route "/rating" is the Url that I and Khalid communicate with the server 
it's the same like `url: "/api/helpful"`, it's our argument

NateWr replied:
I mean specifically the line `.then(response => response.json())` but I think I understand now. That's converting the response string into an object, yeah?

NateWr commented on routes/index.js:
```diff
@@ -25,10 +23,30 @@ router.get('/', function(req, res, next) {
+  const renderEntries = function(error, entry) {
+    if (error) {
+      console.log(error);
+      res.sendStatus(400)
```
A `400` status usually indicates an authentication error (ie - "You are not allowed to do this."). In your case, I think you want to render a `500` status (ie - "The server ran into an error.").

NateWr commented on routes/index.js:
```diff
@@ -25,10 +23,30 @@ router.get('/', function(req, res, next) {
+      {$inc: { [helpful ? 'helpful' : 'unhelpful' ]:  1}},
+      {new: true}
+    , renderEntries);
+});
```
Looks like you might have some indentation issues here. Try formatting the document.

NateWr commented on views/index.hbs:
```diff
@@ -8,25 +8,25 @@
+			{{{this.answer}}}
+		</div>
+		<div class="entry__help">
+			<span class="value-helpful" data-question-id={{this._id}}>{{this.helpful}} people found this helpful. Was this entry helpful?</span>
```
Missing quotes `"`  around `{{this._id}}`.

Also, you're missing the total count: 11/25 people found this helpful. You may need @khaledkzy to modify the data delivered to the template for you.

Yjohn replied:
Khalid accept only the ID and the boolean of the Helpful buttons, and he changed depends on what he gets from me
if it's false he add the total number and if it's true he adds 1 to the helpful

NateWr replied:
Yes, but in this initial template, the string displays without a total count. You only say "11 people found this helpful" but you need to display "11/25 people found this helpful". Yeah?

NateWr commented on public/main.js:
```diff
@@ -0,0 +1,32 @@
+const clickEvent = function (event) {
+    const id = event.target.getAttribute('data-question-id');
+    const helpful = JSON.parse(event.target.getAttribute('data-helpful'));
```
Do you need the `JSON.parse()` here?

Yjohn replied:
yes, I need JSON.parse() to change the string to boolean

NateWr replied:
:+1:

NateWr reviewed (changes requested):
Great work @Yjohn and @khaledkzy! Code looks really clean. Just a few comments throughout. I haven't tested this myself but let me know when you address the comments and I'll give it a go.

Yjohn commented:
@NateWr I commented all the feedback, let me know ... :)

NateWr commented:
@Yjohn Great! You can push new commits to this branch with changes for any outstanding issues (like [this one](https://github.com/NateWr/cyf-faq-group/pull/15#discussion_r139102202)).

Yjohn committed:
format the document

Yjohn committed:
#6 display the total number of People

Yjohn committed:
change the variable name for total entiries

Yjohn committed:
#6 adding loading animation when the button clicked

NateWr commented on views/index.hbs:
```diff
@@ -8,25 +8,26 @@
+			<span class="value-helpful" data-question-id={{this._id}}>{{this.helpful}}/{{this.total}} people found this helpful. Was this entry helpful?</span>
+			<button class="button button--small button-feedback" data-helpful="true" data-question-id="{{this._id}}">Yes</button>
+			<button class="button button--small button-feedback" data-helpful="false" data-question-id="{{this._id}}">No</button>
+			<div id="loading" class="hidden"></div>
```
The `id` attribute needs to be unique across the entire page. So if you have lots of elements with `id="loading"`, it can cause problems if someone tries to use `document.getElementById('loading')`.

It works in your case because you're limiting the selection `document.querySelector('.class-<id> #loading')`. But it should generally be avoided. You can either:

1. Add a unique value to each id: `id="loading-{this._id}"`.

2. Turn the selector into a class: `class="hidden loading".

It's up to you.

Yjohn replied:
I created an id only to apply the animation and style. If someone tries to use `document.getElementById('loading')`, only select empty `<div>` with animations.
I tried to make it `class="hidden loading"` but isn't working

NateWr replied:
Show me the code that you've tried to get it working with a `class`.

Yjohn replied:
# afb3936

NateWr commented on public/main.js:
```diff
@@ -0,0 +1,53 @@
+        let valueElId = valueEl.getAttribute('data-question-id')
+        if (res._id === valueElId) {
+            let totalNumber = res.helpful + res.unhelpful;
+            valueEl.innerHTML = res.helpful + '/' + totalNumber + ' people found this helpful. was this entry helpful';
```
`Was this entry helpful?`

Yjohn replied:
This is how I received changes from the database and server, then display the number of people who select the button of helpful and the total number of people whose give their votes.
which part makes you confused?

NateWr replied:
Sorry, I should have explained. I meant to change the words at the end of the line from `was this entry helpful` to `Was this entry helpful?`

NateWr reviewed (changes requested):
Nice work! The code looks really good. I haven't tested it yet but let me know when you make the last few changes requested and I'll take it for a spin.

Yjohn committed:
change the attributes to be displayed

Yjohn committed:
#6 change the loading ID to the class

NateWr commented:
This is amazing @Yjohn! I love the loading animation too. I've merged it now. :+1:
